PROTOCOL.md

```diff
@@ -2,179 +2,287 @@
+A `channel` is an object which allows 2 concurrent programs to communicate with
+each other. The semantics of a libchan channel are very similar (but not
+identical) to those of Go's native channels. A channel may be used
+synchronously, but do not support Go's channel select semantics.
 
-A channel has 2 ends: a `Sender` end and a `Receiver` end. The Sender can send messages and close the channel.
-The Receiver can receive messages. Messages arrive in the same order they were sent.
+A channel has 2 ends: a `Sender` end and a `Receiver` end. The Sender can send
+messages and close the channel. The Receiver can receive messages. Messages
+arrive in the same order they were sent.
 
-A channel is uni-directional: messages can only flow in one direction. So channels are more similar to pipes
-than to sockets.
+A channel is uni-directional: messages can only flow in one direction. So
+channels are more similar to pipes than to sockets.
 
 ### Message
 
-A message is a discrete packet of data which can be sent on a channel. Messages are structured into multiple
-fields. The protocol defines which data types can be carried by a message, and how transports should encode and
-decode them.
+A message is a discrete packet of data which can be sent on a channel. The
+protocol defines which data types can be carried sent as a message, and how
+transports should encode and decode them.
 
 ### Byte stream
 
-A byte stream is an object which implements raw IO with `Read`, `Write` and `Close` methods.
-Typical byte streams are text files, network sockets, memory buffers, pipes, and so on.
+A byte stream is an object which implements raw IO with `Read`, `Write` and
+`Close` methods. Typical byte streams are text files, network sockets, memory
+buffers, pipes, and so on. A byte stream may either be read only, write only, or
+full duplex.
 
-One distinct characteristic of libchan is that it can encode byte streams as first class fields
-in a message, alongside more basic types like integers or strings.
+One distinct characteristic of libchan is that it can encode byte streams as
+first class fields in a message, alongside more basic types like integers or
+strings.
 
 ### Nesting
 
-Libchan supports nesting. This means that a libchan message can include a channel, which itself
-can be used to send and receive messages, and so on all the way down.
+Libchan supports nesting. This means that a libchan message can include a
+channel, which itself can be used to send and receive messages, and so on all
+the way down.
 
 Nesting is a fundamental property of libchan.
 
 ## Underlying transport
 
-The libchan protocol requires a reliable, 2-way byte stream as a transport.
-The most popular options are TCP connections, unix stream sockets and TLS sessions. 
+The libchan protocol requires a reliable, 2-way byte stream with support for
+multiplexing as a transport. The underlying byte stream protocol is abstracted
```

Comment:
I would add "Whenever this cannot be guaranteed by a transport, a simplicistic multiplexing protocol is definited in LINK, and it can be reused on top of any 2-way byte stream".